In [2]:
#Import necessary libraries
import pandas as pd
import numpy as np
import re # For regular expressions- helps to clean messy text
import nltk # Natural language Toolkit - handles text processing
# Word tokenizer and stopwords come from nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

# Download necessary data (we need it only once)
nltk.download('punkt') # Needed for work_tokenize
nltk.download('stopwords') # Contains list of common words to remove (like 'is'm 'the')

[nltk_data] Downloading package punkt to C:\Users\Aman
[nltk_data]     Sinha/nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.
[nltk_data] Downloading package stopwords to C:\Users\Aman
[nltk_data]     Sinha/nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


True

# Load the dataset

In [3]:
df = pd.read_csv(r"C:\Users\Aman Sinha\Downloads\Datafiniti_Amazon_Consumer_Reviews_of_Amazon_Products_May19.csv") # Dataset link - https://www.kaggle.com/datasets/datafiniti/consumer-reviews-of-amazon-products
df.head()

id             dateAdded           dateUpdated  \
0  AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   
1  AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   
2  AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   
3  AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   
4  AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   

                                                name                  asins  \
0  AmazonBasics AAA Performance Alkaline Batterie...  B00QWO9P0O,B00LH3DMUO   
1  AmazonBasics AAA Performance Alkaline Batterie...  B00QWO9P0O,B00LH3DMUO   
2  AmazonBasics AAA Performance Alkaline Batterie...  B00QWO9P0O,B00LH3DMUO   
3  AmazonBasics AAA Performance Alkaline Batterie...  B00QWO9P0O,B00LH3DMUO   
4  AmazonBasics AAA Performance Alkaline Batterie...  B00QWO9P0O,B00LH3DMUO   

          brand                                         categories  \
0  Amazonbasics  AA,AAA,Health,Electronics,Health & Household,C...   
1  Amazonbasics  AA,AAA,Health,Electronics,Health & Household,C...   
2  Amazonbasics  AA,AAA,Health,Electronics,Health & Household,C...   
3  Amazonbasics  AA,AAA,Health,Electronics,Health & Household,C...   
4  Amazonbasics  AA,AAA,Health,Electronics,Health & Household,C...   

  primaryCategories                                          imageURLs  \
0   Health & Beauty  https://images-na.ssl-images-amazon.com/images...   
1   Health & Beauty  https://images-na.ssl-images-amazon.com/images...   
2   Health & Beauty  https://images-na.ssl-images-amazon.com/images...   
3   Health & Beauty  https://images-na.ssl-images-amazon.com/images...   
4   Health & Beauty  https://images-na.ssl-images-amazon.com/images...   

                                                keys  ... reviews.didPurchase  \
0  amazonbasics/hl002619,amazonbasicsaaaperforman...  ...                 NaN   
1  amazonbasics/hl002619,amazonbasicsaaaperforman...  ...                 NaN   
2  amazonbasics/hl002619,amazonbasicsaaaperforman...  ...                 NaN   
3  amazonbasics/hl002619,amazonbasicsaaaperforman...  ...                 NaN   
4  amazonbasics/hl002619,amazonbasicsaaaperforman...  ...                 NaN   

  reviews.doRecommend reviews.id reviews.numHelpful reviews.rating  \
0                 NaN        NaN                NaN              3   
1                 NaN        NaN                NaN              4   
2                 NaN        NaN                NaN              5   
3                 NaN        NaN                NaN              5   
4                 NaN        NaN                NaN              5   

                                  reviews.sourceURLs  \
0  https://www.amazon.com/product-reviews/B00QWO9...   
1  https://www.amazon.com/product-reviews/B00QWO9...   
2  https://www.amazon.com/product-reviews/B00QWO9...   
3  https://www.amazon.com/product-reviews/B00QWO9...   
4  https://www.amazon.com/product-reviews/B00QWO9...   

                                        reviews.text  \
0  I order 3 of them and one of the item is bad q...   
1  Bulk is always the less expensive way to go fo...   
2  Well they are not Duracell but for the price i...   
3  Seem to work as well as name brand batteries a...   
4  These batteries are very long lasting the pric...   

                                       reviews.title  reviews.username  \
0  ... 3 of them and one of the item is bad quali...        Byger yang   
1  ... always the less expensive way to go for pr...              ByMG   
2  ... are not Duracell but for the price i am ha...  BySharon Lambert   
3  ... as well as name brand batteries at a much ...     Bymark sexson   
4  ... batteries are very long lasting the price ...           Bylinda   

                                          sourceURLs  
0  https://www.barcodable.com/upc/841710106442,ht...  
1  https://www.barcodable.com/upc/841710106442,ht...  
2  https://www.barcodable.com/upc/841710106442,ht...  
3  htt

## Clean the text

In [7]:
# funtion to clean and preprocess each review
def clean_text(text):
    text = text.lower() # Converting all letters in lowercase
    text= re.sub(r'[^a-zA-Z]', ' ', text) # remove everything except alphabets
    tokens= word_tokenize(text) # split sentence into words
    stop_words = set(stopwords.words('english')) # getting the list of common useless words
    cleaned= [word for word in tokens if word not in stop_words] # keeping only meaningful words
    return ' ' .join(cleaned) # join cleaned words back into a string

In [10]:
# Applying cleaning funtion to every review
df['cleaned_text'] = df ['reviews.text'].astype(str).apply(clean_text)

# Show cleaned text
df[['reviews.text','cleaned_text']].head()

reviews.text  \
0  I order 3 of them and one of the item is bad q...   
1  Bulk is always the less expensive way to go fo...   
2  Well they are not Duracell but for the price i...   
3  Seem to work as well as name brand batteries a...   
4  These batteries are very long lasting the pric...   

                                        cleaned_text  
0  order one item bad quality missing backup spri...  
1    bulk always less expensive way go products like  
2                          well duracell price happy  
3  seem work well name brand batteries much bette...  
4                 batteries long lasting price great

## Convert Text into Numbers

In [11]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer() # Initialize count vectorizer: it converts text to numbers
X = vectorizer.fit_transform(df['cleaned_text']) # Fit and tranform the cleanded text into numbers
X.shape # Lets see the shape (rows=reviews. columns = unique words)

(28332, 9790)

# Prepare Labes (Y)

In [15]:
# Convert review reating into sentiment label : 1 for positive and 0 for negative
df['label']= df['reviews.rating'].apply(lambda x: 1 if x>3 else 0)

# extracting target column
y = df['label']


## Train ML model

In [18]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

# Split data into training and test
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.2, random_state=42)

# Initialize Naive Bayes Classifier (Its good for text classification)
model= MultinomialNB()

#Train the model
model.fit(X_train,y_train)

# Predict on test data
y_pred = model.predict(X_test)

# Evaluating the accuracy
print("Accuracy:", accuracy_score(y_test,y_pred))

Accuracy: 0.9181224633845068


I implemented a text classification system using TF-IDF vectorization and Logistic Regression. I achieved an accuracy of ~91.8% on unseen test data. 